MuonAnalysis/MuonAssociators/plugins/MuonFSRProducer.cc

```diff
@@ -0,0 +1,229 @@
+    desc.add<double>("deltaROverEt2Max",0.05)->setComment("max ratio of deltsR(mu,photon) over et2 of the photon");
+    desc.add<double>("isolation",2.0)->setComment("relative isolation cut");
+
+    descriptions.add("MuonFSRProducer", desc);
+  } 
+  ~MuonFSRProducer() override {}
+  
+  //static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+  
+private:
+  
+  void produce(edm::StreamID, edm::Event&, const edm::EventSetup&) const override;
+
+  double computeRelativeIsolation(const pat::PackedCandidate & photon,
+				  const pat::PackedCandidateCollection& pfcands,
+				  const double & isoConeMax,
+				  const double & isoConeMin) const;
+
+// ----------member data ---------------------------
+  const edm::EDGetTokenT<pat::PackedCandidateCollection> pfcands_;
+  const edm::EDGetTokenT<pat::ElectronCollection> electrons_;
+  const edm::EDGetTokenT<edm::View<reco::Muon>> muons_;
+  float ptCut;
+  float etaCut;
+  float photonPtCut;
+  float drEtCut;
+  float isoCut;
+
+
+};
+
+void MuonFSRProducer::produce(edm::StreamID streamID, edm::Event& iEvent, const edm::EventSetup& iSetup) const {
+
+  using namespace std;
+
+  edm::Handle<pat::PackedCandidateCollection> pfcands;
+  iEvent.getByToken(pfcands_, pfcands);
+  edm::Handle<edm::View<reco::Muon>> muons;
+  iEvent.getByToken(muons_, muons);
+  edm::Handle<pat::ElectronCollection> electrons;
+  iEvent.getByToken(electrons_, electrons);
+
+  std::vector<int> muonMapping(muons->size(),-1);
+  auto fsrPhotons = std::make_unique<std::vector<pat::GenericParticle>>();
+  // loop over all muons
+  for (auto muon = muons->begin(); muon != muons->end(); ++muon){
+
+    
+    int photonPosition = -1;
+    double distance_metric_min = -1;
+    // minimum muon pT
+    if (muon->pt() < ptCut) continue;
+    // maximum muon eta
+    if (fabs(muon->eta() > etaCut)) continue;
+
+    // for each muon, loop over all pf cadidates
+    for(auto iter_pf = pfcands->begin(); iter_pf != pfcands->end(); iter_pf++){
+      auto const & pc = *iter_pf;
+
+      
+      // consider only photons
+      if (abs(pc.pdgId()) != 22) continue;
+      //cout<<"ID pass!"<<endl;
+
+      // minimum pT cut
+      if (pc.pt() < photonPtCut) continue;
+      //cout<<"pT pass!"<<endl;
+
+      // eta requirements
+      // if (fabs(pc.eta()) > 1.4442 and (fabs(pc.eta()) < 1.566)) continue;
+      if (fabs(pc.eta()) > 1.4 and (fabs(pc.eta()) < 1.6)) continue;
+      if (fabs(pc.eta()) > 2.5) continue;
+      //cout<<"ETA pass!"<<endl;
+
+      // 0.0001 < DeltaR(photon,muon) < 0.5 requirement
+      double dRPhoMu = deltaR(muon->eta(),muon->phi(),pc.eta(),pc.phi());
+      if(dRPhoMu < 0.0001) continue;
+      if(dRPhoMu > 0.5) continue;
+      //cout<<"DeltaR pass!"<<endl;
+	 
+      bool skipPhoton = false;
+      bool closest = true;
+
+      for (auto othermuon = muons->begin(); othermuon != muons->end(); ++othermuon){
+      	if (othermuon->pt() < ptCut or fabs(othermuon->eta() > etaCut)) continue;
```

Comment:
or othermuon==muon ? (logically, in practice there is strictly less below so I guess it works)

Reply:
added below and othermuon!=muon